agent/acs/handler/payload_responder.go

```diff
@@ -0,0 +1,289 @@
+	"github.com/aws/amazon-ecs-agent/ecs-agent/acs/model/ecsacs"
+	apieni "github.com/aws/amazon-ecs-agent/ecs-agent/api/eni"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/credentials"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	loggerfield "github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/aws-sdk-go/aws"
+)
+
+// skipAddTaskComparatorFunc defines the function pointer that accepts task status
+// and returns the boolean comparison result.
+type skipAddTaskComparatorFunc func(apitaskstatus.TaskStatus) bool
+
+// taskMessageHandler implements TaskMessageHandler.
+type taskMessageHandler struct {
+	taskEngine                  engine.TaskEngine
+	ecsClient                   api.ECSClient
+	dataClient                  data.Client
+	taskHandler                 *eventhandler.TaskHandler
+	credentialsManager          credentials.Manager
+	latestSeqNumberTaskManifest *int64
+}
+
+func (tmHandler *taskMessageHandler) ProcessTaskMessage(message *ecsacs.PayloadMessage,
+	ackFunc func(*ecsacs.AckRequest, []*ecsacs.IAMRoleCredentialsAckRequest)) error {
+
+	credentialsAcks, allTasksHandled := tmHandler.addPayloadTasks(message)
+
+	// Update latestSeqNumberTaskManifest for it to get updated in state file.
+	if tmHandler.latestSeqNumberTaskManifest != nil && message.SeqNum != nil &&
+		*tmHandler.latestSeqNumberTaskManifest < *message.SeqNum {
+		*tmHandler.latestSeqNumberTaskManifest = *message.SeqNum
+	}
+
+	if !allTasksHandled {
+		return fmt.Errorf("did not handle all tasks")
+	}
+
+	// Send ACKs - do it in async such that it does not block handling more tasks.
+	go ackFunc(&ecsacs.AckRequest{
+		Cluster:           message.ClusterArn,
+		ContainerInstance: message.ContainerInstanceArn,
+		MessageId:         message.MessageId,
+	}, credentialsAcks)
+
+	return nil
+}
+
+// addPayloadTasks does validation on each task and, for all valid ones, adds
+// it to the task engine. It returns a bool indicating if it could add every
+// task to the taskEngine and a slice of credential ack requests.
+func (tmHandler *taskMessageHandler) addPayloadTasks(payload *ecsacs.PayloadMessage) (
+	[]*ecsacs.IAMRoleCredentialsAckRequest, bool) {
+	// Verify that we were able to work with all tasks in this payload.
+	// This is so we know whether to ACK the whole thing or not.
+	allTasksOK := true
+
+	validTasks := make([]*apitask.Task, 0, len(payload.Tasks))
+	for _, task := range payload.Tasks {
+		if task == nil {
+			logger.Critical("Received nil task for message", logger.Fields{
+				loggerfield.MessageID: aws.StringValue(payload.MessageId),
+			})
+			allTasksOK = false
+			continue
+		}
+		apiTask, err := apitask.TaskFromACS(task, payload)
+		if err != nil {
+			tmHandler.handleUnrecognizedTask(task, err, payload)
+			allTasksOK = false
+			continue
+		}
+
+		logger.Info("Received task payload from ACS", logger.Fields{
+			loggerfield.TaskARN:       apiTask.Arn,
+			"version":                 apiTask.Version,
```

Comment:
Nit: Maybe worth creating a logger field as "version" seems to be a common field that might be being used across the project.

Reply:
Good point! I will be sure to add this in the next revision if any blocking changes need to be made.

Reply:
Sounds good. Maybe add it in TODOs for the PR for the future reference. Thanks.